akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala

```diff
@@ -3,70 +3,485 @@
+        .map(identity)
+        .addAttributes(Attributes.name("composite-graph"))
+        .toMat(Sink.head)(Keep.left)
+        .run()
+
+      // most specific still the original as the attribute was added on the composite source
+      attributes.get[Name] should contain(Name("original-name"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("composite-graph"))
+    }
+
+    "make the attributes on Sink.fromGraph source behave the same as the stage itself" in {
+      val attributes =
+        Source.maybe.toMat(
+          Sink.fromGraph(new AttributesSink(Attributes.name("original-name")))
+            .withAttributes(Attributes.name("replaced")) // this actually replaces now
+        )(Keep.right)
+          .withAttributes(Attributes.name("whole-graph"))
+          .run()
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "use the initial attributes for dispatcher" in {
+      val dispatcher =
+        Source.fromGraph(new ThreadNameSnitchingStage("my-dispatcher"))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use an explicit attribute on the stage to select dispatcher" in {
+      val dispatcher =
+        Source.fromGraph(
+          // directly on stage
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher")
+            .addAttributes(ActorAttributes.dispatcher("my-dispatcher")))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the most specific dispatcher when another one is defined on a surrounding composed graph" in {
+      val dispatcher =
+        Source.fromGraph(
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher"))
+          .map(identity)
+          // this is now for the composed source -> flow graph
+          .addAttributes(ActorAttributes.dispatcher("my-dispatcher"))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-akka.stream.default-blocking-io-dispatcher")
+    }
+
+    "not change dispatcher from one defined on a surrounding graph" in {
+      val dispatcher =
+        Source.fromGraph(
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher"))
+          // this already introduces an async boundary here
+          .map(identity)
+          // this is now just for map since there already is one inbetween stage and map
+          .async // potential sugar .async("my-dispatcher")
+          .addAttributes(ActorAttributes.dispatcher("my-dispatcher"))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-akka.stream.default-blocking-io-dispatcher")
```

Comment:
As mentioned before, this behaviour is a bit surprising because of the attributes being set on the composite and then not beeing applied to the stage. Here it is possible to follow the code and deduce the outcome. But if a composite comes from a library, then the effect of setting a dispatcher attribute is unknown.